File: smoke_tests/csharp/authorization.cs
// Authorization vulnerabilities in C#
using System;
using System.Web.Mvc;

namespace VulnerableApp
{
    public class AuthorizationController : Controller
    {
        // Test 1: Missing authorization attribute
        public ActionResult AdminDashboard()
        {
            // VULNERABLE: No [Authorize] attribute
            return View();
        }

        // Test 2: Insecure Direct Object Reference (IDOR)
        public ActionResult ViewDocument(int documentId)
        {
            // VULNERABLE: No ownership check
            var document = GetDocument(documentId);
            return Json(document);
        }

        // Test 3: Horizontal privilege escalation
        public ActionResult ViewProfile(int userId)
        {
            // VULNERABLE: Can view any user's profile
            var profile = GetUserProfile(userId);
            return Json(profile);
        }

        // Test 4: Vertical privilege escalation
        [Authorize]
        public ActionResult DeleteUser(int userId)
        {
            // VULNERABLE: No role check for admin action
            DeleteUserById(userId);
            return Ok();
        }

        // Test 5: Client-side authorization
        public ActionResult GetSecretData()
        {
            // VULNERABLE: Relying on client-side check
            // JavaScript checks role but server doesn't
            return Json(new { secret = "sensitive data" });
        }

        // Test 6: Predictable resource IDs
        public ActionResult GetOrder(int orderId)
        {
            // VULNERABLE: Sequential IDs allow enumeration
            var order = GetOrderById(orderId);
            return Json(order);
        }

        // Test 7: Missing function level access control
        public ActionResult ExecuteAdminFunction(string function)
        {
            // VULNERABLE: No check if user can execute function
            ExecuteFunction(function);
            return Ok();
        }

        // Test 8: Path-based authorization bypass
        public ActionResult AdminApi()
        {
            string path = Request.Path;
            // VULNERABLE: Path check can be bypassed
            // /ADMIN/api or /admin/../admin/api
            if (path.ToLower().StartsWith("/admin"))
            {
                return Json(new { data = "admin data" });
            }
            return Unauthorized();
        }

        // Test 9: JWT without signature verification
        public ActionResult ValidateToken()
        {
            string token = Request.Headers["Authorization"];
            // VULNERABLE: Not verifying JWT signature
            var parts = token.Split('.');
            var payload = parts[1];  // Just decoding, not verifying
            return Ok();
        }

        // Test 10: Role check in wrong place
        public ActionResult UpdateSettings()
        {
            UpdateSystemSettings();  // Action happens first
            // VULNERABLE: Check after action
            if (!User.IsInRole("Admin"))
            {
                return Unauthorized();
            }
            return Ok();
        }

        // Test 11: Trusting user-provided role
        public ActionResult ActionWithRole()
        {
            string role = Request.Headers["X-User-Role"];
            // VULNERABLE: Trusting client-provided role
            if (role == "admin")
            {
                return Json(new { adminData = true });
            }
            return Ok();
        }

        // Test 12: Cached authorization
        [OutputCache(Duration = 3600)]
        [Authorize(Roles = "Admin")]
        public ActionResult CachedAdminPage()
        {
            // VULNERABLE: Cached page may be served to non-admins
            return View();
        }

        // Placeholder methods
        private object GetDocument(int id) => new { };
        private object GetUserProfile(int id) => new { };
        private void DeleteUserById(int id) { }
        private object GetOrderById(int id) => new { };
        private void ExecuteFunction(string f) { }
        private void UpdateSystemSettings() { }
    }
}

File: smoke_tests/php/authorization.php
<?php
// Authorization vulnerabilities in PHP

// Test 1: Missing authorization check
function admin_dashboard() {
    // VULNERABLE: No authorization check
    $users = get_all_users();
    return render_dashboard($users);
}

// Test 2: IDOR - Insecure Direct Object Reference
function view_document() {
    $document_id = $_GET['id'];
    // VULNERABLE: No ownership check
    $document = get_document($document_id);
    return render_document($document);
}

// Test 3: Horizontal privilege escalation
function view_profile() {
    $user_id = $_GET['user_id'];
    // VULNERABLE: Can view any user's profile
    $profile = get_user_profile($user_id);
    return render_profile($profile);
}

// Test 4: Vertical privilege escalation
function delete_user() {
    $user_id = $_POST['user_id'];
    // VULNERABLE: No admin check
    delete_user_by_id($user_id);
    return redirect('/users');
}

// Test 5: Client-side authorization
function get_secret_data() {
    // VULNERABLE: Relying on JavaScript for authorization
    return json_encode(['secret' => 'sensitive data']);
}

// Test 6: Predictable resource IDs
function get_order() {
    $order_id = $_GET['id'];
    // VULNERABLE: Sequential IDs allow enumeration
    $order = get_order_by_id($order_id);
    echo json_encode($order);
}

// Test 7: Missing function level access control
function execute_function() {
    $function = $_GET['function'];
    // VULNERABLE: No permission check
    if (function_exists($function)) {
        return $function();
    }
}

// Test 8: Path-based authorization bypass
function admin_api() {
    $path = $_SERVER['REQUEST_URI'];
    // VULNERABLE: Can bypass with case or encoding
    if (stripos($path, '/admin') === 0) {
        return json_encode(['data' => 'admin data']);
    }
    http_response_code(401);
}

// Test 9: Check after action
function update_settings() {
    $settings = $_POST['settings'];
    // VULNERABLE: Action happens before authorization
    update_system_settings($settings);

    if (!is_admin()) {
        http_response_code(401);
        return;
    }
    return redirect('/settings');
}

// Test 10: Trusting user-provided role
function action_with_role() {
    $role = $_SERVER['HTTP_X_USER_ROLE'];
    // VULNERABLE: Trusting client header
    if ($role === 'admin') {
        return json_encode(['admin_data' => true]);
    }
    http_response_code(403);
}

// Test 11: Cached authorization
function cached_admin_page() {
    // VULNERABLE: Cached page served to non-admins
    header('Cache-Control: public, max-age=3600');
    return render_admin_page();
}

// Test 12: Incomplete authorization chain
function edit_post() {
    $post_id = $_POST['post_id'];
    $content = $_POST['content'];

    // VULNERABLE: Only checks if logged in, not ownership
    if (is_logged_in()) {
        update_post($post_id, $content);
        return redirect('/posts/' . $post_id);
    }
    http_response_code(401);
}

// Helper functions
function get_all_users() { return []; }
function render_dashboard($users) {}
function get_document($id) { return []; }
function render_document($doc) {}
function get_user_profile($id) { return []; }
function render_profile($profile) {}
function delete_user_by_id($id) {}
function redirect($url) { header("Location: $url"); }
function get_order_by_id($id) { return []; }
function update_system_settings($settings) {}
function is_admin() { return false; }
function is_logged_in() { return true; }
function render_admin_page() {}
function update_post($id, $content) {}
?>
